File: DSAlgorithms/Leetcode/ts/src/arrays_hash/118-pascal_triangle.ts
const generate = (numRows: number): number[][] =>{
    const triangle: number[][] = [];
    const first_row: number[] = [1];
    triangle.push(first_row);
    for(let i = 1; i < numRows; i++)
    {
        const previous_row: number[] = triangle[i - 1];
        const current_row:number[] = [1];
        for(let j = 1; j < i; j++)
            current_row.push(previous_row[j - 1] + previous_row[j]);
        current_row.push(1);
        triangle.push(current_row);
    }
    return triangle;
};

File: dsa_algo/leetcode/arrays/arrays_hash/118-pascal-triangle.cpp
vector<vector<int>> generate(int numRows) {
    std::vector<std::vector<int>> triangle;
    for(int i = 0; i < numRows; i++) 
    {
        std::vector<int> row(i + 1, 1);
        for(int j = 1; j < i; j++)
            row[j] = triangle[i - 1][j] + triangle[i - 1][j - 1];
        triangle.push_back(row);
    }
    return triangle;
}
vector<vector<int>> pascal_triangle(int numRows) {
    std::vector<std::vector<int>> triangle;
    if(numRows == 0) 
        return triangle;
    std::vector<int> first_row;
    first_row.push_back(1);
    triangle.push_back(first_row);
    for(int i =1; i < numRows; i++)
    {
        std::vector<int> prev_row = triangle[i - 1];
        std::vector<int> current_row;
        current_row.push_back(1);
        for(int j = 1; j < i; j++)
            current_row.push_back(prev_row[j - 1] + prev_row[j]);
        current_row.push_back(1);
        triangle.push_back(current_row);
    }
    return triangle;
}
